snowdrop-execution/src/Snowdrop/Execution/DbActions/Types.hs

```diff
@@ -4,35 +4,105 @@
+    -- ^ Modify change accumulator with specified list of change sets.
+    -- First argument is a change accumulator to modify.
+    -- Second argument is a list of change sets (in oldest first order).
+    -- Function applies each change set to accumulator in oldest first order,
+    -- returns modified accumulator.
+    }
+
+-- | Actions to execute 'DbAccessU' effect.
+-- Methods provided in the data type are not intended to modify any internal state.
+data DbAccessActionsU chgAccum undo id value m = DbAccessActionsU
+    { daaAccessM          :: DbAccessActionsM chgAccum id value m
+    -- ^ Actions to execute 'DbAccessM' effect.
+    , daaModifyAccumUndo  :: chgAccum
+                          -> NewestFirst [] undo
+                          -> m (Either (CSMappendException id) chgAccum)
+    -- ^ Applies undos to a given accumulator.
+    -- First argument is a change accumulator to modify.
+    -- Second argument is a list of undo object (in newest first order).
+    -- Function applies each undo to accumulator in newest first order,
+    -- returns modified accumulator.
+    , daaComputeUndo      :: chgAccum
+                          -> chgAccum
+                          -> m (Either (CSMappendException id) undo)
+    -- ^ Computes undo. First argument is a base change accumulator.
+    -- Second argument is a modified change accumulator.
+    -- Function returns an undo object, such that @modified `applyUndo` undo = base@
     }
 
--- | Db modify access actions model following use of
-data DbModifyActions chgAccum id value m proof = DbModifyActions
-    { dmaAccessActions :: DbAccessActions chgAccum id value m
-    , dmaApply         :: chgAccum -> m proof
+-- | Class provided to execute effect with ations corresponding to it.
+class DbActions effect actions param m where
+    -- | Execute @effect r@ using @actions m@ and return value @r@ in monad @m@.
+    executeEffect :: effect r -> actions m -> param -> m r
+
+instance Functor m => DbActions (DbAccess id value)
+                                (DbAccessActions chgAccum id value) chgAccum m where
+    executeEffect (DbQuery req cont) daa chgAccum =
+        cont <$> daaGetter daa chgAccum req
+    executeEffect (DbIterator prefix (FoldF (e, acc, cont))) daa chgAccum =
+        cont <$> daaIter daa chgAccum prefix e acc
+
+instance Functor m => DbActions (DbAccessU chgAccum undo id value)
+                                (DbAccessActionsU chgAccum undo id value) chgAccum m where
+    executeEffect (DbAccessM daM) (daaAccessM -> daaM) chgAccum = executeEffect daM daaM chgAccum
+    executeEffect (DbComputeUndo cs cont) daaU chgAccum =
+        cont <$> daaComputeUndo daaU chgAccum cs
+    executeEffect (DbModifyAccumUndo undos cont) daaU chgAccum =
+        cont <$> daaModifyAccumUndo daaU chgAccum undos
+
+instance Functor m => DbActions (DbAccessM chgAccum id value)
+                                (DbAccessActionsM chgAccum id value) chgAccum m where
```

Comment:
May `DbActions  DbAccessM ... ` instance be moved before  `DbActions DbAccessU ...` instance? We have this order on other places.